### ISSUE TITLE
[HOLD for payment 2023-09-29] [$1000] Chat - Compose box becomes overlapped by the keyboard when sroll to the begginig

### ISSUE THREAD

**Issue by lanitochka17** [Expensify/App#23986]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
Precondition: user should be logged in and have converation with mesages
1. Open https://staging.new.expensify.com/
2. Tap on any listed conversation
3. Tap in the compose box
4. Scroll to beginning of the conversation
5. At the beginning, pull the top of the conversation down, as if to refresh
6. Scroll back down

## Expected Result:
Compose box should be properly displayed

## Actual Result:
Compose box becomes overlapped by the keayobard when scroll to the beggining

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [x] iOS / Safari
- [ ] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.3.48.0

**Reproducible in staging?:** Yes

**Reproducible in production?:** Yes

**If this was caught during regression testing, add the test name, ID and link from TestRail:**

**Email or phone of affected tester (no customers):**

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/78819774/2f653a6f-0824-4a01-8a4a-ffb3280f363a

**Expensify/Expensify Issue URL:**

**Issue reported by:** Applause- Internal Team

**Slack conversation:**

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~0135c5fbbbebd34a7a</li>
        <li>Upwork Job ID: 1686574863342690304</li>
        <li>Last Price Increase: 2023-08-09</li>
    </ul>
</details>

**Comment by jscardona12** [comment]
## Proposal
### Please re-state the problem that we are trying to solve in this issue.
Chat - Compose box becomes overlapped by the keyboard when sroll to the begginig
### What is the root cause of that problem?
This is a known issue in safari when the keyboard is open the user is able to scroll thowugh the window messing up with the UI.
### What changes do you think we should make in order to solve the problem?
We should set up a window scroll listener in 
in [Composer/index.js](https://github.com/Expensify/App/blob/a97a15476d9dfccc5e2b0ed8e05b9803d5f8e4c4/src/components/Composer/index.js#L157) in componentDidMount we should add the following lines to detect when keyboar is being open on browser iOS:
                 
            window.addEventListener("scroll", () => {
            // if is the first time the window scrolls dave the scrollY to detect if is the right amount of windowScroll
            if(this.state.windowScrollY === -1){
                this.setState({windowScrollY: window.scrollY})
            }
            // if there is a scroll and its not 0(keyboard is hidden) or this.state.windowScrollY (KeyBoard is shown) prevent scroll
            else if(window.scrollY !== 0 && window.scrollY !== this.state.windowScrollY && this.textInput.isFocused()){
               //Prevent Window Scroll
                this.state.windowScrollY !== -1 ? window.scrollTo(0,this.state.windowScrollY) : null;
            }
        });



### Without Fix


https://github.com/Expensify/App/assets/16781411/9d10de64-6210-44e7-aec7-109486bdeb6e



###  With Fix

https://github.com/Expensify/App/assets/16781411/09da287a-c274-4d7b-a092-a053587ca61b

**Comment by sophiepintoraetz** [comment]
Can dupe:
![IMG_1765](https://github.com/Expensify/App/assets/42940078/b7ff0b82-7c0b-4762-8eb9-dc5dff726918)

**Comment by mananjadhav** [comment]
@sophiepintoraetz Can you confirm if the expected behavior should keyboard should be closed while scrolling down? or if the composer + keyboard should be visible?

**Comment by Crystalian0322** [comment]
Proposal
Please re-state the problem that we are trying to solve in this issue.
Chat - Compose box becomes overlapped by the keyboard when sroll to the begginig

What is the root cause of that problem?
This is a known issue in safari when the keyboard is open the user is able to scroll thowugh the window messing up with the UI.

What changes do you think we should make in order to solve the problem?
We should set up a window scroll listener in
in [Composer/index.js](https://github.com/Expensify/App/blob/a97a15476d9dfccc5e2b0ed8e05b9803d5f8e4c4/src/components/Composer/index.js#L157) in componentDidMount we should add the following lines to detect when keyboar is being open on browser iOS:

        window.addEventListener("scroll", () => {
        // if is the first time the window scrolls dave the scrollY to detect if is the right amount of windowScroll
        if(this.state.windowScrollY === -1){
            this.setState({windowScrollY: window.scrollY})
        }
        // if there is a scroll and its not 0(keyboard is hidden) or this.state.windowScrollY (KeyBoard is shown) prevent scroll
        else if(window.scrollY !== 0 && window.scrollY !== this.state.windowScrollY && this.textInput.isFocused()){
           //Prevent Window Scroll
            this.state.windowScrollY !== -1 ? window.scrollTo(0,this.state.windowScrollY) : null;
        }
    });

**Comment by studentofcoding** [comment]
# Proposal

## Please re-state the problem that we are trying to solve in this issue
Compose box Chat becomes overlapped with the keyboard when scrolling to the top

## What is the root cause of that problem?
As the other shared before, this is the familiar problem where the user can scroll through a window thus, the Compose Box become overlapped

## What changes do you think we should make in order to solve the problem?
What we need to do is create `handleScroll()` function below `getNextChars` and add an Event listener on  `componentDidMount()` to handle `handleScroll()`
```
window.addEventListener("scroll", () => {
            this.handleScroll();
        });
...
handleScroll() {
        // Check if the windowScrollY property is undefined
        if (this.windowScrollY === undefined) {
            // If it is undefined, set it to the current scroll position of the window
            this.windowScrollY = window.scrollY;
        }
        // Check if the windowScrollY property is not equal to the current scroll position of the window
        if (this.windowScrollY === -1) {
            // If it is not equal, set it to the current scroll position of the window
            this.windowScrollY = window.scrollY;
        }
        // if there is a scroll and its not 0(keyboard is hidden) or this.state.windowScrollY (KeyBoard is shown) prevent scroll
        if (window.scrollY !== 0 && window.scrollY !== this.windowScrollY) {
            // If the conditions are met, scroll the window back to the stored windowScrollY position
            window.scrollTo(0, this.windowScrollY);
        }
    }
```

With this solution, we can always see the Compose box appear on the screen, whether it's on Safari or Chrome

## What alternative solutions did you explore? (Optional)
NA

##Result
- [x] On IOS/Safari

https://github.com/Expensify/App/assets/20473526/8353790e-c81a-4dd2-a5cc-f5f50c36afad


- [x] On IOS/Chrome

https://github.com/Expensify/App/assets/20473526/3f13faa8-864c-4071-8507-5800604f8177

cc: @sophiepintoraetz

**Comment by sophiepintoraetz** [comment]
My expectation would be that if the cursor is active in the composer box, the keyboard would remain open, right? @mananjadhav

**Comment by mananjadhav** [comment]
Thanks for the confirmation. I think we should put this on hold until we merge https://github.com/Expensify/App/pull/23359.

The PR refactors the Composer/index.js to a functional component, which would require changes in the approach here.

**Comment by kosmydel** [comment]
Hey, I am from the Software Mansion expert's agency. I know it is on hold, but my changes don't affect the refactor. Here is my proposal:

## Proposal
### Please re-state the problem that we are trying to solve in this issue.
When the user scrolls to the beginning of the report and keeps scrolling, the keyboard overflows the input bar.

### What is the root cause of that problem?
When the user scrolls to the beginning of the report, the browser allows the user to continue scrolling, which leads to scrolling the container (the entire page). 

### What changes do you think we should make in order to solve the problem?
In my opinion, the above solutions overcomplicate this issue. I believe, that the only thing we have to do is to add this style to the [`FlatList`](https://github.com/Expensify/App/blob/dd0d98fa6868bba60e81bfdf66e5cdf7d2adaa37/src/components/InvertedFlatList/BaseInvertedFlatList.js#L131):
```javascript
style={{overscrollBehavior: 'contain'}}
```

This styling will block [(docs)](https://developer.mozilla.org/en-US/docs/Web/CSS/overscroll-behavior) the scrolling when the user comes to the beginning of the report.

**Demo**

https://github.com/Expensify/App/assets/104823336/ba29cc21-98d5-4df3-8e3a-f2a7c7d900cb


### What alternative solutions did you explore? (Optional)
I don't think that we need to overcomplicate it.

**Comment by mananjadhav** [comment]
Okay while the other issue is still in progress, I can @kosmydel's proposal looks good. I will have to test the scroll behavior on all cases, but based on the docs it should work. Also it shouldn't get blocked by the refactor.

**Comment by mananjadhav** [comment]
🎀 👀 🎀

**Comment by studentofcoding** [comment]
> Okay while the other issue is still in progress, I can @kosmydel's proposal looks good. I will have to test the scroll behavior on all cases, but based on the docs it should work. Also it shouldn't get blocked by the refactor.
> 
> 
> 
> cc - @sophiepintoraetz @neil-marcellini 
> 
> 
> 
> 🎀 👀 🎀 C+ reviewed.

Hey @mananjadhav do the order of proposal matter in this case?

**Comment by studentofcoding** [comment]
The reason is on `CONTRIBUTING.md`

https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#for-external-agencies-that-expensify-partners-with

We have this condition of

```
- If no proposals have been submitted by other contributors, BugZero (BZ) team member or an internal engineer will assign the issue to you.
- **If there are existing proposals, BZ will put the issue on hold. [Contributor+]**(https://github.com/Expensify/App/blob/main/contributingGuides/HOW_TO_BECOME_A_CONTRIBUTOR_PLUS.md) will review the existing proposals. If a contributor’s proposal is accepted then the contributor will be assigned to the issue. If not the issue will be assigned to the agency-employee.
```

Thus in this case the review process should prioritize Contributor first then Agency next

**Comment by mananjadhav** [comment]
@studentofcoding None of the existing proposals were good enough to solve it. @kosmydel's proposal was better, simple and fixes the issue.

Also I would strongly advise not to copy other's proposal and edit in yours. 

<img width="1174" alt="image" src="https://github.com/Expensify/App/assets/3069065/3cc2a1e7-426b-476a-91e4-de3627a372a6">

**Comment by mananjadhav** [comment]
@kosmydel I would recommend avoid opening the PR, until you're officially assigned the job. Let the internal engineer (in this case @neil-marcellini), to confirm and assign you the issue.

**Comment by kosmydel** [comment]
Hey @neil-marcellini , can I ask for the assignment?

**Comment by neil-marcellini** [comment]
I'll review today

**Comment by neil-marcellini** [comment]
Wow, great [proposal](https://github.com/Expensify/App/issues/23986#issuecomment-1674543447) @kosmydel! You're hired.

**Comment by kosmydel** [comment]
Hey, thanks!
I've prepared [a PR](https://github.com/Expensify/App/pull/24502) for that. Do you want me to add this style to other `FlatList`'s as well?

**Comment by neil-marcellini** [comment]
If you're confident it won't break anything else, than sure. Otherwise let's keep the scope limited.

**Comment by mananjadhav** [comment]
I would recommend keep the scope limited, let it go through staging QA, and then create a follow up issue.

**Comment by sophiepintoraetz** [comment]
All right:
- Applause reported
- @mananjadhav - please raise a manual request for $1000 once you have completed the checklist
- @kosmydel is from agency so no payments here.

**Comment by JmillsExpensify** [comment]
$1,000 payment approved for @mananjadhav based on BZ summary.
